pydantic_ai_slim/pydantic_ai/ag_ui.py

```diff
@@ -0,0 +1,871 @@
+                yield ToolCallArgsEvent(
+                    type=EventType.TOOL_CALL_ARGS,
+                    tool_call_id=agent_event.delta.tool_call_id
+                    or stream_ctx.last_tool_call_id
+                    or 'unknown',  # Should never be unknown, but just in case.
+                    delta=agent_event.delta.args_delta
+                    if isinstance(agent_event.delta.args_delta, str)
+                    else json.dumps(agent_event.delta.args_delta),
+                )
+            elif isinstance(agent_event.delta, ThinkingPartDelta):  # pragma: no cover
+                yield ThinkingTextMessageContentEvent(
+                    type=EventType.THINKING_TEXT_MESSAGE_CONTENT,
+                    delta=agent_event.delta.content_delta or '',
+                )
+        elif isinstance(agent_event, FinalResultEvent):
+            # No equivalent AG-UI event yet.
+            pass
+
+
+@dataclass
+class _History:
+    """A simple history representation for AG-UI protocol."""
+
+    prompt_message_id: str  # The ID of the last user message.
+    messages: list[ModelMessage]
+
+
+def _convert_history(messages: list[Message]) -> _History:
+    """Convert a AG-UI history to a PydanticAI one.
+
+    Args:
+        messages: List of AG-UI messages to convert.
+
+    Returns:
+        List of PydanticAI model messages.
+    """
+    msg: Message
+    prompt_message_id: str = ''
+    result: list[ModelMessage] = []
+    tool_calls: dict[str, str] = {}  # Tool call ID to tool name mapping.
+    for msg in messages:
+        if isinstance(msg, UserMessage):
+            prompt_message_id = msg.id
+            result.append(ModelRequest(parts=[UserPromptPart(content=msg.content)]))
+        elif isinstance(msg, AssistantMessage):
+            if msg.tool_calls:
+                for tool_call in msg.tool_calls:
+                    tool_calls[tool_call.id] = tool_call.function.name
+
+                result.append(
+                    ModelResponse(
+                        parts=[
+                            ToolCallPart(
+                                tool_name=tool_call.function.name,
+                                tool_call_id=tool_call.id,
+                                args=tool_call.function.arguments,
+                            )
+                            for tool_call in msg.tool_calls
+                        ]
+                    )
+                )
+
+            if msg.content:
+                result.append(ModelResponse(parts=[TextPart(content=msg.content)]))
+        elif isinstance(msg, SystemMessage):
+            result.append(ModelRequest(parts=[SystemPromptPart(content=msg.content)]))
+        elif isinstance(msg, ToolMessage):
+            result.append(
+                ModelRequest(
+                    parts=[
+                        ToolReturnPart(
+                            tool_name=tool_calls.get(msg.tool_call_id, 'unknown'),
```

Comment:
If the tool call ID is not in `tool_calls`, something has gone wrong! So I'd prefer to add an assert or some exception, and drop the `unknown`